pyrefly/lib/lsp/server.rs

```diff
@@ -190,6 +190,43 @@ use crate::state::semantic_tokens::SemanticTokensLegends;
 use crate::state::state::State;
 use crate::state::state::Transaction;
 
+/// Interface exposed for outside modules to interact with the LSP server
+pub trait LspServerInterface {
+    /// Send a response back to the LSP client
+    fn send_response(&self, response: Response);
+
+    /// Process an LSP event and return the next step
+    fn process_event<'a>(
+        &'a self,
+        ide_transaction_manager: &mut TransactionManager<'a>,
+        canceled_requests: &mut HashSet<RequestId>,
+        subsequent_mutation: bool,
+        event: LspEvent,
+    ) -> anyhow::Result<ProcessEvent>;
+}
+
+/// Factory for creating LSP server instances that implement LspServerInterface
+pub struct LspServerFactory;
+
+impl LspServerFactory {
+    /// Create a new LSP server instance
+    pub fn create(
+        connection: Arc<Connection>,
+        lsp_queue: LspQueue,
+        initialization_params: InitializeParams,
+        indexing_mode: IndexingMode,
+        workspace_indexing_limit: usize,
+    ) -> Box<dyn LspServerInterface> {
+        Box::new(Server::new(
+            connection,
+            lsp_queue,
+            initialization_params,
+            indexing_mode,
+            workspace_indexing_limit,
+        ))
+    }
+}
+
```

Comment:
I think you can skip the box if you skip stack allocation below and likely don't need the factory